src/domain/siwe/siwe.repository.ts

```diff
@@ -106,4 +110,37 @@ export class SiweRepository implements ISiweRepository {
       await this.siweApi.clearNonce(message.nonce);
     }
   }
+
+  /**
+   * Verifies signature of signed SiWe message, either by EOA or smart contract
+   *
+   * @param args.message - SiWe message
+   * @param args.chainId - chainId of the blockchain
+   * @param args.address - address of the signer
+   * @param args.signature - signature from signing {@link args.message}
+   *
+   * @returns boolean - whether the signature is valid
+   */
+  private async isValidSignature(args: {
+    message: string;
+    chainId: string;
+    address: `0x${string}`;
+    signature: `0x${string}`;
+  }): Promise<boolean> {
+    // First check if valid signature of EOA as it can be done off chain
+    const isValidEoaSignature = await verifyMessage(args).catch(() => false);
+    if (isValidEoaSignature) {
+      return true;
+    }
+
+    // Else, verify hash on-chain using ERC-6492 for smart contract accounts
+    const blockchainApi = await this.blockchainApiManager.getBlockchainApi(
+      args.chainId,
+    );
+    const client = blockchainApi.getClient();
+    return client.verifySiweMessage({
+      message: args.message,
+      signature: args.signature,
+    });
```

Comment:
Related with this other comment from the previous PR: https://github.com/safe-global/safe-client-gateway/pull/1623#discussion_r1634962430

Maybe it's simpler if we create and expose `blockchainApi.verifySiweMessage`. For the testing/mocking part it would simplify things also IMO, what do you think?